Feat/setup contracts bonding curve

Zitzak opened the pull request:
## Thoughts about the structure
- Most base contract (BondingCurveFundingManagerBase) implements functions concerning the issuing (buying) of tokens, which every bonding curve funding manager must have. It also defines an abstract function `_issueTokensFormulaWrapper` which must be used in downstream contracts to define through which formula price is calculated
- The next base contract (RedeemingBondingCurveFundingMangerBase) implements the functions concerning the redemption of tokens back for their collateral. It also defines an abstract function `_redeemTokensFormulaWrapper` which must be used in downstream contracts to define through which formula price is calculated. The reason to have two of these functions is because the Bancor contract has a different interface for calculating issuing tokens vs redeeming tokens
- There are 2 more abstract contracts which introduce a virtual issuing token supply (VirtualTokenSupplyBase) and a virtual collateral supply (VirtualCollateralSupplyBase). The would be considered even more specialised in terms of functionality and therefor would be added more in a compositing fashion instead of strict inheritence
- The final contract has all the features above and is called BancorVirtualSupplyBondingCurveFundingManger

Zitzak committed:
feat(BondingCurveFundingManagerBase): Add base features

The base contract enables the basic functionalitiy of issuing token
which every bonding curve must have. It makes use of a abstract function
for the calculation which should be implemented in downstream contracts

Zitzak committed:
feat(RedeemingBondingCurveFundingManager): Add redeem functionalities

This base contract adds configuration function for a redeeming bonding
curve functionality as well as abastract function for calculating
token values to be implemented in an downstream contract

Zitzak committed:
feat(VirtualSupply): implements virtual token and collateral supply

Zitzak committed:
feat(BancorVirtualSupplyBondingCurveFundingManger):

Inherits from the base contracts to enable issuing and redeeming
of tokens on a bonding curve, calculated through the Bancor formula,
based on virtual supplies of the issuing token and collateral

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/formula/SafeMath.sol:
I think safemath isnt needed anymore since compiler version 0.8 so you can probably remove this and refactor the bancor Formular accordingly

0xNuggan replied:
I'm not sure we should touch the Bancor Formula if not necessary...

Zitzak replied:
Ended up adding the files back to the repo, see conversation on [Discord](https://discord.com/channels/788422482087641089/1006521951445078117/1141012690505039943)

I have added a detailed description of the changes I made within each contract, as well as added the files to the ignore list for the `fmt`, so it wouldn't completely change the file

I will document these changes within our documentation

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/formula/BancorUtils.sol:
This can just directly be included in the Bancor formular contract in my opinion.

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/formula/BancorFormula.sol:
Where exactly did you get this from?
Are there any tests for it?

FHieser replied:
Do we want to do either:
Import it into lib
or integrate it into the Repository directly (+ Tests)

0xNuggan replied:
Yes, maybe we can just import this as a library? What do you think @Zitzak  ?

Zitzak replied:
See first comment

FHieser commented on src/modules/fundingManager/token/IVirtualSupplyToken.sol:
Do these token contracts serve any purpose currently?

FHieser commented on src/modules/fundingManager/token/VirtualSupplyBase.sol:
Do these token contracts serve any purpose currently?

FHieser commented on src/modules/fundingManager/token/VirtualSupplyTokenUpgradeable.sol:
Do these token contracts serve any purpose currently?

FHieser commented on test/modules/fundingManager/BondingCurveFundingManager.t.sol:
Outcommented

FHieser commented on test/modules/fundingManager/PrimaryMarketMaker.t.sol:
Outcommented

FHieser commented on test/modules/fundingManager/token/SetVirtualSupply.t.sol:
Outcommented

FHieser commented on test/modules/fundingManager/token/VirtualSupplyTokenDeployment.t.sol:
Outcommented

FHieser commented on test/modules/fundingManager/token/VirtualSupplyTokenTest.sol:
Outcommented

FHieser commented on test/modules/fundingManager/token/utils/mocks/VirtualSupplyTokenUpgradeableMock.sol:
Outcommented

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/marketMaker/IVirtualTokenSupply.sol:
```diff
@@ -0,0 +1,7 @@
+
+pragma solidity 0.8.19;
+
+interface IVirtualTokenSupply {
```
I would add the getters here too

Zitzak replied:
Done

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/marketMaker/IVirtualCollateralSupply.sol:
```diff
@@ -0,0 +1,7 @@
+
+pragma solidity 0.8.19;
+
+interface IVirtualCollateralSupply {
```
I would add the getters here too

FHieser replied:
If you want to Force the usage of modifiers in these functions you can instead create virtual/abstract functions
maybe in that regard you want to use super.setVirtualCollateralSupply()

Zitzak replied:
Done

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/BondingCurveFundingManagerBase.sol:
```diff
@@ -0,0 +1,143 @@
+        return tokenDecimals;
+    }
+
+    function deposit(uint amount) external {}
```
Can you remind me to bring this up in the upcoming Devmeeting that these functions are not necessary in the IFundingManager Interface every time

0xNuggan replied:
Yeah, let's discuss this in the meeting. I may be ready to see the light myself in that regard.

FHieser commented on src/modules/fundingManager/bondingCurveFundingManager/BondingCurveFundingManagerBase.sol:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    function _setCollateral(address _collateral) internal {
+        collateral = _collateral;
```
Isnt the collateral just the address(this).balance?
Wasnt that the main point of the Virtual Collateral to be able to be set?

Zitzak replied:
Could you explain this point again for me please. You are right that the collateral address is just the token of defined by the Orchestrator contract, However we talked about also the future possibility to add other collateral tokens in our first call, which is why I included this setter function.

I'm unsure what is meant by balance

FHieser replied:
Forget what I said here 🙈
This is to set the underlying payment token address right?
Because I apparently cant read I assumed that this would somehow set a "virtual" Collateral Amount here, which it straight up doesnt

FHieser reviewed (commented):
All in all looks good

Zitzak committed:
feat:BancorVirtualSupplyBondingCurveFundingManager

Remove Virtual Supply token and collateral interfaces

Zitzak committed:
feat: add getter func and errors to interfaces

The getter function got added as per PR review request.  Next to this
errors got added to the interface for over/underflow when setting
the virtual supply

Zitzak committed:
feat: add setter func to VirtualSupply contract

The setter function has been implemented as abstract
function so the upstream contract has to implement it.
Because the getter functions got added to the interfaces,
both contract now implement the interface. Lastly some
input validation got added

Zitzak committed:
test: virtual supply token contracts

Zitzak committed:
feat: add Bancor contracts to repo

All contracts have added comments for clarification
what has been changed, including links to the
original files.

An ignore for the formula directory got added to the
foundry.toml file to prevent the formatter from creating
to big of a diff, in case somebody would like to check our
statement of the changes made.

Zitzak committed:
chore: update modifier after rebase

FHieser commented on remappings.txt:
```diff
@@ -1,2 +1,2 @@
 @oz-up/=lib/openzeppelin-contracts-upgradeable/contracts/
-@oz/=lib/openzeppelin-contracts/contracts/
+@oz/=lib/openzeppelin-contracts/contracts/
```
lol what is happening here?
it added it and then removed it again?

0xNuggan commented on src/modules/fundingManager/bondingCurveFundingManager/BancorVirtualSupplyBondingCurveFundingManager.sol:
```diff
@@ -0,0 +1,158 @@
+        override(RedeemingBondingCurveFundingManagerBase)
+    {
+        // WiP
+        // Q: Deduct fee from token or collateral?
```
When selling, deducting the fee from the collateral after the token has been burned simplifies the process: you use the math as intended and end up with collected fees in only one token.

Zitzak replied:
Thank you for the feedback. Deducting fee will be implemented in a different PR

0xNuggan commented on src/modules/fundingManager/bondingCurveFundingManager/BondingCurveFundingManagerBase.sol:
```diff
@@ -0,0 +1,143 @@
+    bool internal buyIsOpen;
+    uint internal buyFee;
+    address internal collateral;
+    // Q: Do we want to set other beneficiary in the contract then the module itself for the fees?
```
I think transferOrchestratorToken() is good enough to move the fees out when the time comes (if the fees are going to go to the worklflow owners anyway). We'd probably need a way to track the collected amount, though

Zitzak replied:
I agree and will not be implementing a beneficiary on top of the standard way to pull out tokens

0xNuggan commented on src/modules/fundingManager/bondingCurveFundingManager/BondingCurveFundingManagerBase.sol:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    function _setTokenDecimals(uint8 _decimals) internal {
+        // Input check
```
What's the thought behind this? This would be the decimals of the token we mint and give out, right? Is there a reason we wouldn't want to stick with 18?

Zitzak replied:
I brought this up in a call with Felix. He suggested that I could stick to what was implemented before and follow the same pattern. To be more specific, the RebaseFundingManager also enables the setting of decimals so I decided to do the same

0xNuggan replied:
Ok yeah, makes sense to keep it consistent

0xNuggan commented on src/modules/fundingManager/bondingCurveFundingManager/BondingCurveFundingManagerBase.sol:
```diff
@@ -0,0 +1,143 @@
+        external
+        onlyOrchestrator
+    {
+        __Module_orchestrator.token().safeTransfer(to, amount);
```
Since we are going to remove orchestrator.token() anyway, maybe go ahead and use the collateral address directly?

Zitzak replied:
So the initial idea was to add a collateral address to the contract. The reason I discussed with Felix was because in the future we might want to make use of multiple collateral tokens.

I have however removed the reference to the collateral token in the code I'm currently writing (next PR) because setting up the contract for a future with multiple collateral tokens makes me assume to many things. I decided instead to go with what we need right now which means that the code should work with one collateral token, which is the one defined by the Orchestrator.

For this reason, I would leave the transfer of the collateral like this, and update once we remove the `orchestrator.token()` to however it will be implemented once this is needed.

Happy to hear your thoughts around this? :)

0xNuggan replied:
Yeah, I think we can keep it like that then. Once we remove `orchestrator.token()` we can refactor it, but not get ahead of ourselves until then. 
Having multiple tokens seems a change big enough to wait for the rest to be in place. Also I think it's out of the scope of the BCRG doc?

FHieser committed:
Merge branch 'main' into feat/setup_contracts_bonding_curve
